# Welcome to Python

In [1]:
my_name = "Alex"
print(my_name)

Alex


# Lesson 1: Variables, Types, and Operators

*This tutorial was generated from a Jupyter notebook.  You can download the notebook [here](l04_more_operators_and_conditionals.ipynb).*

## Learning Objectives
In this lesson, we will learn about **variables**, a critical concept for programming in Python.  We will be able to determine and manipulate the **type** of a variable and gain a thorough understanding of the importance of these types and their differences. Finally, we will use **operators** to work with our variables to achieve a desired output. To facilitate learning, we will make use of the Python native **print** function to visualize outputs from our code.

## What is a variable?
A **variable** is a Python object that can be defined by the user. All variables have 2 important properties:
1. The **type** of the variable
2. The **value** of the variable

### Assigning a Variable
Let's start by creating, or **instantiating** a variable. To do so, we use the assignment operator, `=`, where the left side is the variable and the right side is the value.

In [ ]:
my_name = 'John'

We just created our first variable! But how do we check to make sure our variable is what we expect?
The `print()` function can be used to output the value of variables.

In [2]:
print(my_name)

Alex


Let's create and view
some more variables.

In [7]:
num1 = 10
num2 = 5.2
num3 = 2

In [8]:
print(num1)

10


In [9]:
print(num2)

5.2


In [10]:
print(num3)

2


The `print()` function also allows us to print multiple variables in one line!

In [11]:
print(num1, num2, num3)

10 5.2 2


When variables are created, they are stored in **memory**. For now, just remember that a variable does not exist until it is instantiated, and instantiated variables are stored in memory for reference until they are cleared from memory.

Also, variable names cannot contain spaces. Often, **underscore** `_` is used to separate words in variable names.

## Variable Types
As stated, all variables have a **type** and a **value**. The type of a variable determines what you can do with it and, as such, is extremely important. We can easily see what the type of a variable is by using the Python native `type()` function.

In [12]:
print(type(my_name))

<class 'str'>


In [13]:
print(type(num1))

<class 'int'>


In [14]:
print(type(num2))

<class 'float'>


In [15]:
print(type(num3))

<class 'int'>


We can see that our defined variables have different types! Let's talk about 3 basic types of variables:
- `int`: Integer, or whole number, that is positive or negative and without decimals
- `float`: Floating point number that is a decimal or fractional value (NOT an integer)
- `str`: Sequence of characters, where each character is uniquely identified by Python (Space is a character!)

### More On Strings
You may have noticed that we assigned the value **'John'** to the variable **my_name**, but the `print()` output of my_name is **John**. Strings are created by enclosing characters in quotes. To demonstrate, let's try to instantiate another string variable.

In [16]:
my_new_name = "Billy"

As you can see, we were given an error! Without the quotes, Python interprets **Billy** as a variable, but we have not defined a variable with that name (it is not in the memory!). Let's correctly instantiate our new string.

In [17]:
my_full_name = 'John Doe'

In [18]:
print(my_full_name)

John Doe


**Note**: While we will exclusively use **single quotes `'`**, strings can be instantiated using **double quotes `"`** or **triple quotes (`'''`, `"""`)**. Triple quotes allow for extension of strings over multiple lines which can be useful for keeping track of long strings.

## Comments
As a programmer, it is unlikely you will perfectly remember what every line of your code does. In addition, it is impractical to explain your code to other potential users by just showing them the code. In Python, **comments** are lines of code that are ignored by the interpreter. They are often used to add information to your code that makes it more readable and interpretable. It is great practice to write code that is well-commented. Your future self will thank you!

Comments are denoted by the `#` symbol. In a single line, any text after the "#" is part of the comment. This allows for an entire line to be a comment, or a comment to be place beside a line of code to be executed. Here's an example of how comments can be used to add information for understanding:

In [19]:
# I am going to print all of the variables that I have instantiated so far.

In [20]:
print(my_name)
print(my_full_name) # This variable is my full name!
print(num1)
print(num2)
print(num3)

Alex
John Doe
10
5.2
2


Another effective usage of comments is to comment a line of code you currently do not want to be executed. This is extremely helpful for debugging! Here is an example:

In [21]:
# I do not want to print my first name, only my full name
#print(my_name)
print(my_full_name) # This variable is my full name!
print(num1)
print(num2)
print(num3)

John Doe
10
5.2
2


Finally, while not shown here, **triple quotes** `'''` can be used to create comments that go over multiple lines. This is useful for noting large amounts important information in your script.

## Operators
Operators allow you to do things with variables, like add them. They are represented by special symbols, like `+` and `*`. For now, we will focus on Python's arithmetic operators, as denoted below:
|Action|Operator|
|:-------|:----------:|
|Addition | `+`|
|Subtraction | `-`|
|Multiplication | `*`|
|Division | `/`|
|Raise to Power | `**`|
|Modulo | `%`|
|Floor Division | `//`|

### Operations on Integers
Let’s see how these operators work on integers.

In [22]:
2 + 3 # Addition

5

In [23]:
2 - 3 # Subtraction

-1

In [24]:
2 * 3 # Multiplication

6

In [25]:
2 / 3 # Division

0.6666666666666666

In [26]:
2 ** 3 # Raise to Power

8

In [27]:
2 % 3 # Modulo

2

In [28]:
2 // 3 # Floor Division

0

### Operations on Floats
Let's try floats.

In [29]:
2.1 + 3.2

5.300000000000001

**Wait a minute!** We know the result should be 5.3, but Python gives 5.300000000000001. This is due to the fact that floating point numbers are stored with a finite number of binary bits. There will always be some rounding errors. This means that as far as the computer is concerned, it cannot tell you that 2.1 + 3.2 and 5.3 are equal. This is important to remember when dealing with floats, as we will see in the next lesson.

In [30]:
2.1 - 3.2

-1.1

In [31]:
# Very close to zero because of finite precision
5.3 - (2.1 + 3.2)

-8.881784197001252e-16

In [32]:
2.1 * 3.2

6.720000000000001

In [33]:
2.1 / 3.2

0.65625

In [34]:
2.1 ** 3.2

10.74241047739471

In [35]:
2.1 % 3.2

2.1

In [36]:
2.1 // 3.2

0.0

Note that with `int`s or `float`s, division by 0 is not possible.

In [37]:
2.1 / 0.0

ZeroDivisionError: ignored

In [38]:
2 / 0

ZeroDivisionError: ignored

### Operations on Integers and Floats
This proceeds as expected.

In [39]:
2.1 + 3

5.1

In [40]:
2.1 - 3

-0.8999999999999999

In [41]:
2.1 * 3

6.300000000000001

In [42]:
2.1 / 3

0.7000000000000001

In [43]:
2.1 ** 3

9.261000000000001

In [44]:
2.1 % 3

2.1

In [45]:
2.1 // 3

0.0

### Operations on Strings
Now let’s try some of these operations on strings. This idea of applying mathematical operations to strings seems strange, but let’s just mess around and see what we get.

In [46]:
'Hello, ' + 'world.'

'Hello, world.'

Ah! Adding strings together concatenates them! This is also intuitive. How about subtracting strings?

In [48]:
'Hello, ' - 'world.'

TypeError: ignored

That stands to reason. Subtracting strings does not make sense. The Python interpreter was kind enough to give us a nice error message saying that we can’t have a `str` and a `str` operand type for the subtraction operation. It also makes sense that we can’t do multiplication, raising of power, etc., with two strings. How about multiplying a string by an integer?

In [49]:
'Hello, world.' * 3

'Hello, world.Hello, world.Hello, world.'

Yes, this makes sense! Multiplication by an integer is the same thing as just adding multiple times, so the Python interpreter concatenates the string several times.

As a final note on operators with strings, watch out for this:

In [50]:
'4' + '2'

'42'

The result is not `6`, but it is a string containing the characters `4` and `2`.

### Order of Operations
The order of operations is also as we would expect. Exponentiation comes first, followed by multiplication and division, floor division, and modulo. Next comes addition and subtraction. In order of precedence, our arithmetic operator table is
|Precedence|Operators|
|:-------|:----------:|
|1 | `**`|
|2 | `*`, `/`, `//`, `%`|
|3 | `+`. `-`|

You can also group operations with parentheses. Operations within parentheses are always evaluated first. Be careful to not use excessive parentheses, as this can lead to complex code with bugs that are difficult to find.

Let’s practice.

In [51]:
1 + 4**2

17

In [52]:
1 + 4/2

3.0

In [53]:
1**3 + 2**3 + 3**3 + 4**3

100

In [54]:
(1 + 2 + 3 + 4)**2

100

Interestingly, we also demonstrated that the sum of the first ***n*** cubes is equal to the sum of the first ***n*** integers squared. Fun!

## Variables and Assignment Operators
So far, we have essentially just used Python as an oversized desktop calculator. We would really like to be able to think about our computational problems symbolically. We learned about variables at the beginning of the tutorial, but in practice we were just using numbers and strings directly. We would like to say that a variable, a, represents an integer and another variable b represents another integer. Then, we could do things like add a and b. We now see the intuition behind variables having to have a type and a value associated with them so the Python interpreter knows what to do when we use operators with them.

Let's instantiate new variables, `a` and `b`, and add them.

In [55]:
a = 2
b = 3
a + b

5

Great! We get what we expect! And we still have `a` and `b`.

In [56]:
print(a, b)

2 3


Now, we might be tempted to say, “`a` is two.” No. `a` is not two. `a` is a variable that has a value of 2. A variable in Python is not just its value. It also has more associated with it under the hood of the interpreter that we will not get into. So, you can think about a variable as a map to an address in RAM (called a pointer in computer-speak) that stores information, including a type and a value.

### Assignment/Increment Operators
Now, let’s say we wanted to update the value of `a` by adding 4.1 to it. Python will do some magic for us.

In [57]:
print(type(a), a)

a = a + 4.1

print(type(a), a)

<class 'int'> 2
<class 'float'> 6.1


We see that `a` was initially an integer with a value of 2. But we added 4.1 to it, so the Python interpreter knew to change its type to a float and update its value.

This operation of updating a value can also be accomplished with an **increment operator**.

In [58]:
a = 2
a += 4.1
print(a)

6.1


The `+=` operator told the interpreter to take the value of `a` and add 4.1 to it, changing the type of `a` in the intuitive way if need be. The other six arithmetic operators have similar constructions, `-=`, `*=`, `/=`, `//=`, `%=`, and `**=`.

In [59]:
a = 2
a **= 3
print(a)

8


## Type Conversion
Suppose you have a variable of one type, and you want to convert it to another. For example, say you have a string, `'42'`, and you want to convert it to an integer. This would happen if you were reading information from a text file, which by definition is full of strings, and you wanted to convert some string to a number. This is done as follows.

In [60]:
my_str = '42'
my_int = int(my_str)
print(my_int, type(my_int))

42 <class 'int'>


Conversely, we can convert an `int` back to a `str`.

When converting a `float` to an `int`, the interpreter does not round the result, but gives the floor.

In [61]:
int(2.9)

2

Also consider our string concatenation warning/example from above:

In [62]:
print('4' + '2')
print(int('4') + int('2'))

42
6
